### PR TITLE
Add QUIC support for SCION NTSKE

### PR DESCRIPTION
**Pull request by aaronbojarski**
Open questions:

- [ ] naming convention for NTSKE server functions (SCION/IP or QUIC/TCP)
- [ ] in timeservice remove QUIC examples
- [ ] default QUIC port

### CODE REVIEW

File: net/ntske/fetcher.go [hunk 1]
```diff
@@ -3,44 +3,86 @@ package ntske
...
-		return errors.New("unexpected NTS-KE meta data: no cookies")
-	}
-	if ke.Meta.Algo != AES_SIV_CMAC_256 {
-		return errors.New("unexpected NTS-KE meta data: unknown algorithm")
-	}
+		err = ExportKeys(conn.ConnectionState().TLS.ConnectionState, &f.data)
+		if err != nil {
+			return err
+		}
+
```
Comment (marcfrei) [inline net/ntske/fetcher.go:56]: Remove empty line.

File: net/ntske/fetcher.go [hunk 2]
```diff
@@ -3,44 +3,86 @@ package ntske
...
+		if len(f.data.Cookie) == 0 {
+			return errNoCookies
+		}
+		if f.data.Algo != AES_SIV_CMAC_256 {
+			return errUnknownAEAD
+		}
```
Comment (marcfrei) [inline net/ntske/fetcher.go:77]: Are these checks not also needed in the QUIC case?

File: timeservice.go [hunk 3]
```diff
@@ -198,22 +199,27 @@ func (c *ntpReferenceClockIP) MeasureClockOffset(ctx context.Context, log *zap.L
...
 		InsecureSkipVerify: ntskeInsecureSkipVerify,
 		ServerName:         ntskeHost,
 		MinVersion:         tls.VersionTLS13,
 	}
 	c.Auth.NTSKEFetcher.Port = ntskePort
 	c.Auth.NTSKEFetcher.Log = log
+	c.Auth.NTSKEFetcher.SCIONQuic.Enabled = true
+	c.Auth.NTSKEFetcher.SCIONQuic.RemoteAddr = remoteAddr
+	c.Auth.NTSKEFetcher.SCIONQuic.LocalAddr = localAddr
+	c.Auth.NTSKEFetcher.SCIONQuic.DaemonAddr = daemonAddr
```
Comment (marcfrei) [inline timeservice.go:219]: Reverse order of assignments.

File: core/server/ntske_scion.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+				log.Info("failed to handle connection",
+					zap.Stringer("remote", conn.RemoteAddr()),
+					zap.Error(err),
+				)
+			}
+		}()
+	}
+}
+
+func StartSCIONNTSKEServer(ctx context.Context, log *zap.Logger, localIP net.IP, localPort int, config *tls.Config, provider *ntske.Provider, localAddr udp.UDPAddr) {
```
[the remark this answers is not in the document]
Comment (marcfrei) [reply core/server/ntske_scion.go:105]: Signature: `(ctx context.Context, log *zap.Logger, localAddr udp.UDPAddr, config *tls.Config, provider *ntske.Provider)`

File: core/server/ntske_scion.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	//ntskeAddr := net.JoinHostPort(localIP.String(), strconv.Itoa(defaultNtskePort))
+	log.Info("server listening via SCION",
+		zap.Stringer("ip", localIP),
```
Comment (marcfrei) [inline core/server/ntske_scion.go:108]: `localAddr.Host.IP`

File: core/server/ntske_scion.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	)
+
+	localAddr.Host.Port = defaultNtskePort
+
+	listener, err := scion.ListenQUIC(ctx, localAddr, config, nil /* quicCfg */)
+	if err != nil {
+		log.Fatal("failed to start listening", zap.Error(err))
+	}
+
+	go runSCIONNTSKEServer(ctx, log, listener, localPort, provider)
```
Comment (marcfrei) [inline core/server/ntske_scion.go:119]: `localAddr.Host.Port`?

File: timeservice.go [hunk 7]
```diff
@@ -460,6 +469,7 @@ func runRelay(configFile string) {
 	provider := ntske.NewProvider()
 
 	server.StartNTSKEServer(ctx, log, copyIP(localAddr.Host.IP), localAddr.Host.Port, tlsConfig, provider)
+	server.StartSCIONNTSKEServer(ctx, log, copyIP(localAddr.Host.IP), localAddr.Host.Port, tlsConfig, provider, udp.UDPAddrFromSnet(localAddr))
```
Comment (marcfrei) [inline timeservice.go:472]: `server.StartNTSKEServerSCION(ctx, log, udp.UDPAddrFromSnet(localAddr), tlsConfig, provider)`

**Review by marcfrei — CHANGES_REQUESTED** [review]

File: timeservice.go [hunk 8]
```diff
@@ -198,22 +199,27 @@ func (c *ntpReferenceClockIP) MeasureClockOffset(ctx context.Context, log *zap.L
 	return client.MeasureClockOffsetIP(ctx, log, c.ntpc, c.localAddr, c.remoteAddr)
 }
 
-func configureSCIONClientNTS(c *client.SCIONClient, ntskeServer string, ntskeInsecureSkipVerify bool) {
+func configureSCIONClientNTS(c *client.SCIONClient, ntskeServer string, ntskeInsecureSkipVerify bool, remoteAddr, localAddr udp.UDPAddr, daemonAddr string) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply timeservice.go]: Done

File: timeservice.go [hunk 9]
```diff
@@ -198,22 +199,27 @@ func (c *ntpReferenceClockIP) MeasureClockOffset(ctx context.Context, log *zap.L
...
 	c.Auth.NTSKEFetcher.Port = ntskePort
 	c.Auth.NTSKEFetcher.Log = log
+	c.Auth.NTSKEFetcher.SCIONQuic.Enabled = true
+	c.Auth.NTSKEFetcher.SCIONQuic.RemoteAddr = remoteAddr
+	c.Auth.NTSKEFetcher.SCIONQuic.LocalAddr = localAddr
+	c.Auth.NTSKEFetcher.SCIONQuic.DaemonAddr = daemonAddr
 }
 
-func newNTPReferenceClockSCION(localAddr, remoteAddr udp.UDPAddr,
+func newNTPReferenceClockSCION(localAddr, remoteAddr udp.UDPAddr, daemonAddr string,
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply timeservice.go]: Done

File: net/ntske/ntske_scion.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	}
+	quic.SendStream(stream).Close()
+	reader := bufio.NewReader(stream)
+
+	// Wait for response
+	err = Read(log, reader, data)
+	if err != nil {
+		return err
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go:106]: Done

File: net/ntske/ntske_scion.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+func ExchangeQUIC(log *zap.Logger, conn *scion.QUICConnection, data *Data) error {
+	stream, err := conn.OpenStream()
+	if err != nil {
+		return err
+	}
+	defer quic.SendStream(stream).Close()
+
+	var msg ExchangeMsg
+	var nextproto NextProto
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go]: Done

File: net/ntske/ntske_scion.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func ExchangeQUIC(log *zap.Logger, conn *scion.QUICConnection, data *Data) error {
+	stream, err := conn.OpenStream()
+	if err != nil {
+		return err
+	}
+	defer quic.SendStream(stream).Close()
+
+	var msg ExchangeMsg
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go:76]: Done

File: net/ntske/ntske_scion.go [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+	return conn, Data{}, nil
+}
+
+func ExchangeQUIC(log *zap.Logger, conn *scion.QUICConnection, data *Data) error {
+	stream, err := conn.OpenStream()
+	if err != nil {
+		return err
+	}
+	defer quic.SendStream(stream).Close()
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go]: Done

File: net/ntske/ntske_scion.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+	conn, err := scion.DialQUIC(ctx, localAddr, remoteAddr, sp,
+		"" /* host*/, config, nil /* quicCfg */)
+	if err != nil {
+		return nil, Data{}, err
+	}
+
+	return conn, Data{}, nil
+}
+
+func ExchangeQUIC(log *zap.Logger, conn *scion.QUICConnection, data *Data) error {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go]: Done

File: net/ntske/ntske_ip.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	return conn, data, nil
+}
+
+func ExchangeTCP(log *zap.Logger, conn *tls.Conn, data *Data) error {
+	reader := bufio.NewReader(conn)
+
+	var msg ExchangeMsg
+	var nextproto NextProto
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go]: Done

File: net/ntske/ntske_ip.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+		return nil, Data{}, errServerNoNTSKE
+	}
+
+	return conn, data, nil
+}
+
+func ExchangeTCP(log *zap.Logger, conn *tls.Conn, data *Data) error {
+	reader := bufio.NewReader(conn)
+
+	var msg ExchangeMsg
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go:65]: Done

File: net/ntske/ntske_ip.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+	state := conn.ConnectionState()
+	if state.NegotiatedProtocol != alpn {
+		return nil, Data{}, errServerNoNTSKE
+	}
+
+	return conn, data, nil
+}
+
+func ExchangeTCP(log *zap.Logger, conn *tls.Conn, data *Data) error {
+	reader := bufio.NewReader(conn)
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go]: Done

File: net/ntske/ntske_ip.go [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+	state := conn.ConnectionState()
+	if state.NegotiatedProtocol != alpn {
+		return nil, Data{}, errServerNoNTSKE
+	}
+
+	return conn, data, nil
+}
+
+func ExchangeTCP(log *zap.Logger, conn *tls.Conn, data *Data) error {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go]: Done

File: net/ntske/ntske_scion.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+	config.NextProtos = []string{alpn}
+	ctx := context.Background()
+
+	dc := newDaemonConnector(log, ctx, daemonAddr)
+	ps, err := dc.Paths(ctx, remoteAddr.IA, localAddr.IA, daemon.PathReqFlags{Refresh: true})
+	if err != nil {
+		log.Fatal("failed to lookup paths", zap.Stringer("to", remoteAddr.IA), zap.Error(err))
+	}
+	if len(ps) == 0 {
+		log.Fatal("no paths available", zap.Stringer("to", remoteAddr.IA))
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go]: Done

File: net/ntske/ntske_scion.go [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func ConnectQUIC(log *zap.Logger, localAddr, remoteAddr udp.UDPAddr, daemonAddr string, config *tls.Config) (*scion.QUICConnection, Data, error) {
+	config.NextProtos = []string{alpn}
+	ctx := context.Background()
+
+	dc := newDaemonConnector(log, ctx, daemonAddr)
+	ps, err := dc.Paths(ctx, remoteAddr.IA, localAddr.IA, daemon.PathReqFlags{Refresh: true})
+	if err != nil {
+		log.Fatal("failed to lookup paths", zap.Stringer("to", remoteAddr.IA), zap.Error(err))
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go]: Done

File: net/ntske/ntske_scion.go [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+func NewQUICListener(ctx context.Context, listener quic.Listener) (quic.Connection, error) {
+	conn, err := listener.Accept(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+func ConnectQUIC(log *zap.Logger, localAddr, remoteAddr udp.UDPAddr, daemonAddr string, config *tls.Config) (*scion.QUICConnection, Data, error) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go]: Done

File: net/ntske/ntske_ip.go [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+		Timeout: time.Second * 5,
+	}, "tcp", hostport, config)
+	if err != nil {
+		return nil, Data{}, err
+	}
+
+	var data Data
+	data.Server, _, err = net.SplitHostPort(conn.RemoteAddr().String())
+	if err != nil {
+		return nil, Data{}, err
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go:51]: Done

File: net/ntske/ntske_ip.go [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+	tlsConn, ok := conn.(*tls.Conn)
+	if !ok {
+		return nil, errConvertTLS
+	}
+
+	return tlsConn, nil
+}
+
+func ConnectTCP(hostport string, config *tls.Config) (*tls.Conn, Data, error) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go]: Done

File: net/ntske/ntske_ip.go [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+	"errors"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+var errConvertTLS = errors.New("could not convert to tls connection")
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go]: Done

File: net/ntske/ntske_scion.go [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+		Address: daemonAddr,
+	}
+	c, err := s.Connect(ctx)
+	if err != nil {
+		log.Fatal("failed to create demon connector", zap.Error(err))
+	}
+	return c
+}
+
+func NewQUICListener(ctx context.Context, listener quic.Listener) (quic.Connection, error) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go]: Done

File: net/ntske/ntske_ip.go [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+
+func NewTCPListener(listener net.Listener) (*tls.Conn, error) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go]: Done

File: net/ntske/ntske_ip.go [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+	conn, err := listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConn, ok := conn.(*tls.Conn)
+	if !ok {
+		return nil, errConvertTLS
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go]: Done

File: net/ntske/ntske_ip.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_ip.go]: Done

File: net/ntske/ntske.go [hunk 28]
```diff
@@ -71,8 +60,22 @@ const (
 	DEFAULT_NTP_PORT   = 123
 )
 
+const (
+	UnrecognizedCriticalErrorCode = 0
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske.go:64]: Done

File: net/ntske/ntske.go [hunk 29]
```diff
@@ -71,8 +60,22 @@ const (
...
+	BadRequestErrorCode           = 1
+	InternalServerErrorCode       = 2
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske.go:66]: Done

File: net/ntske/ntske.go [hunk 30]
```diff
@@ -71,8 +60,22 @@ const (
...
+)
+
 const alpn = "ntske/1"
 
+var (
+	errServerNoNTSKE            = errors.New("server not speaking ntske/1")
+	errReadInternalServer       = errors.New("ntske received internal server error message")
+	errReadBadRequest           = errors.New("ntske received bad request error message")
+	errReadUnrecognisedCritical = errors.New("ntske received unrecognised critical error message")
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske.go:75]: Done

File: net/ntske/ntske.go [hunk 31]
```diff
@@ -71,8 +60,22 @@ const (
...
+const (
+	UnrecognizedCriticalErrorCode = 0
+	BadRequestErrorCode           = 1
+	InternalServerErrorCode       = 2
+)
+
 const alpn = "ntske/1"
 
+var (
+	errServerNoNTSKE            = errors.New("server not speaking ntske/1")
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske.go:72]: Done

File: core/server/ntske.go [hunk 32]
```diff
@@ -1,35 +1,19 @@
...
 	"net"
-	"strconv"
 
 	"go.uber.org/zap"
 
 	"example.com/scion-time/net/ntske"
 )
 
-const defaultNtskePort int = 4460
+const defaultNtskePort = 4460
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske.go:12]: Done

File: core/server/ntske_scion.go (hunk 5 above)
Comment (aaronbojarski) [reply core/server/ntske_scion.go:108]: Done

File: core/server/ntske_scion.go (hunk 6 above)
Comment (aaronbojarski) [reply core/server/ntske_scion.go:119]: Done

File: timeservice.go [hunk 33]
```diff
@@ -429,6 +437,7 @@ func runServer(configFile string) {
 	provider := ntske.NewProvider()
 
 	server.StartNTSKEServer(ctx, log, copyIP(localAddr.Host.IP), localAddr.Host.Port, tlsConfig, provider)
+	server.StartSCIONNTSKEServer(ctx, log, copyIP(localAddr.Host.IP), localAddr.Host.Port, tlsConfig, provider, udp.UDPAddrFromSnet(localAddr))
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply timeservice.go:440]: Done

File: timeservice.go (hunk 7 above)
Comment (aaronbojarski) [reply timeservice.go:472]: Done

File: core/server/ntske_scion.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_scion.go:105]: Done

File: core/server/ntske_scion.go [hunk 34]
```diff
@@ -0,0 +1,120 @@
...
+	log.Info("server listening via SCION",
+		zap.Stringer("ip", localIP),
+		zap.Int("port", defaultNtskePort),
+	)
+
+	localAddr.Host.Port = defaultNtskePort
+
+	listener, err := scion.ListenQUIC(ctx, localAddr, config, nil /* quicCfg */)
+	if err != nil {
+		log.Fatal("failed to start listening", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_scion.go:116]: Done

File: core/server/ntske_ip.go [hunk 35]
```diff
@@ -0,0 +1,102 @@
...
+func StartNTSKEServer(ctx context.Context, log *zap.Logger, localIP net.IP, localPort int, config *tls.Config, provider *ntske.Provider) {
+	ntskeAddr := net.JoinHostPort(localIP.String(), strconv.Itoa(defaultNtskePort))
+	log.Info("server listening via IP",
+		zap.Stringer("ip", localIP),
+		zap.Int("port", defaultNtskePort),
+	)
+
+	listener, err := tls.Listen("tcp", ntskeAddr, config)
+	if err != nil {
+		log.Error("failed to create TLS listener")
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_ip.go:98]: Done

File: core/server/ntske_scion.go [hunk 36]
```diff
@@ -0,0 +1,120 @@
...
+	if err != nil {
+		return err
+	}
+
+	quic.SendStream(stream).Close()
+
+	return nil
+}
+
+func runSCIONNTSKEServer(ctx context.Context, log *zap.Logger, listener quic.Listener, localPort int, provider *ntske.Provider) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_scion.go:83]: Done

File: core/server/ntske_ip.go [hunk 37]
```diff
@@ -0,0 +1,102 @@
...
+		conn, err := ntske.NewTCPListener(listener)
+		if err != nil {
+			log.Info("failed to accept client", zap.Error(err))
+			continue
+		}
+		go handleTCPKeyExchange(log, conn, localPort, provider)
+	}
+}
+
+func StartNTSKEServer(ctx context.Context, log *zap.Logger, localIP net.IP, localPort int, config *tls.Config, provider *ntske.Provider) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_ip.go:89]: Done

File: core/server/ntske_ip.go [hunk 38]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	n, err := conn.Write(buf.Bytes())
+	if err != nil || n != buf.Len() {
+		log.Info("failed to write response", zap.Error(err))
+		return
+	}
+}
+
+func runNTSKEServer(log *zap.Logger, listener net.Listener, localPort int, provider *ntske.Provider) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_ip.go:77]: Done

File: core/server/ntske_scion.go [hunk 39]
```diff
@@ -0,0 +1,120 @@
...
+		sendNtskeQuicErrorMessage(log, stream, ntske.InternalServerErrorCode)
+		return err
+	}
+
+	_, err = stream.Write(buf.Bytes())
+	if err != nil {
+		return err
+	}
+
+	quic.SendStream(stream).Close()
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_scion.go:78]: Done

File: core/server/ntske_scion.go [hunk 40]
```diff
@@ -0,0 +1,120 @@
...
+	if err != nil {
+		return err
+	}
+	defer quic.SendStream(stream).Close()
+
+	reader := bufio.NewReader(stream)
+	var data ntske.Data
+	err = ntske.Read(log, reader, &data)
+	if err != nil {
+		sendNtskeQuicErrorMessage(log, stream, ntske.BadRequestErrorCode)
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_scion.go]: Done

File: core/server/ntske_scion.go [hunk 41]
```diff
@@ -0,0 +1,120 @@
...
+
+func handleQUICKeyExchange(log *zap.Logger, conn quic.Connection, localPort int, provider *ntske.Provider) error {
+	stream, err := conn.AcceptStream(context.Background())
+	if err != nil {
+		return err
+	}
+	defer quic.SendStream(stream).Close()
+
+	reader := bufio.NewReader(stream)
+	var data ntske.Data
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_scion.go:45]: Done

File: core/server/ntske_scion.go [hunk 42]
```diff
@@ -0,0 +1,120 @@
...
+		return
+	}
+}
+
+func handleQUICKeyExchange(log *zap.Logger, conn quic.Connection, localPort int, provider *ntske.Provider) error {
+	stream, err := conn.AcceptStream(context.Background())
+	if err != nil {
+		return err
+	}
+	defer quic.SendStream(stream).Close()
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_scion.go:42]: Done

File: core/server/ntske_scion.go [hunk 43]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	n, err := stream.Write(buf.Bytes())
+	if err != nil || n != buf.Len() {
+		log.Info("failed to write error message", zap.Error(err))
+		return
+	}
+}
+
+func handleQUICKeyExchange(log *zap.Logger, conn quic.Connection, localPort int, provider *ntske.Provider) error {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_scion.go:37]: Done

File: core/server/ntske_scion.go [hunk 44]
```diff
@@ -0,0 +1,120 @@
...
+
+	"github.com/quic-go/quic-go"
+	"go.uber.org/zap"
+
+	"example.com/scion-time/net/ntske"
+	"example.com/scion-time/net/scion"
+	"example.com/scion-time/net/udp"
+)
+
+func sendNtskeQuicErrorMessage(log *zap.Logger, stream quic.Stream, code int) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_scion.go:18]: Done

File: core/server/ntske_ip.go [hunk 45]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	n, err := conn.Write(buf.Bytes())
+	if err != nil || n != buf.Len() {
+		log.Info("failed to write error message", zap.Error(err))
+		return
+	}
+}
+
+func handleTCPKeyExchange(log *zap.Logger, conn *tls.Conn, localPort int, provider *ntske.Provider) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_ip.go:34]: Done

File: core/server/ntske_ip.go [hunk 46]
```diff
@@ -0,0 +1,102 @@
...
+	"crypto/tls"
+	"net"
+	"strconv"
+
+	"go.uber.org/zap"
+
+	"example.com/scion-time/net/ntske"
+)
+
+func sendNtskeTcpErrorMessage(log *zap.Logger, conn *tls.Conn, code int) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske_ip.go:15]: Done

File: core/server/ntske.go [hunk 47]
```diff
@@ -1,35 +1,19 @@
...
-	err = ke.ExportKeys()
-	if err != nil {
-		log.Info("failed to export keys", zap.Error(err))
-		return
-	}
-
-	localIP := ke.Conn.LocalAddr().(*net.TCPAddr).IP
+var errNoCookie = errors.New("failed to add at least one cookie")
 
+func newNtskeMessage(log *zap.Logger, localIP net.IP, localPort int, data *ntske.Data, provider *ntske.Provider) (ntske.ExchangeMsg, error) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/server/ntske.go:16]: Done

File: net/ntske/fetcher.go (hunk 2 above)
Comment (aaronbojarski) [reply net/ntske/fetcher.go]: Yes done

File: net/ntske/fetcher.go [hunk 48]
```diff
@@ -3,44 +3,86 @@ package ntske
...
-	}
-	if ke.Meta.Algo != AES_SIV_CMAC_256 {
-		return errors.New("unexpected NTS-KE meta data: unknown algorithm")
-	}
+		err = ExportKeys(conn.ConnectionState().TLS.ConnectionState, &f.data)
+		if err != nil {
+			return err
+		}
+
+	} else {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/fetcher.go]: Done

File: net/ntske/fetcher.go (hunk 1 above)
Comment (aaronbojarski) [reply net/ntske/fetcher.go]: Done

File: net/ntske/fetcher.go [hunk 49]
```diff
@@ -3,44 +3,86 @@ package ntske
...
 	"go.uber.org/zap"
+
+	"github.com/quic-go/quic-go"
+
+	"example.com/scion-time/net/udp"
+)
+
+var (
+	errNoCookies   = errors.New("unexpected NTS-KE meta data: no cookies")
+	errUnknownAEAD = errors.New("unexpected NTS-KE meta data: unknown algorithm")
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/fetcher.go]: Done

File: net/ntske/fetcher.go [hunk 50]
```diff
@@ -3,44 +3,86 @@ package ntske
...
 )
 
 type Fetcher struct {
 	Log       *zap.Logger
 	TLSConfig tls.Config
 	Port      string
-	data      Data
+	SCIONQuic struct {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/fetcher.go]: Done

File: net/ntske/fetcher.go [hunk 51]
```diff
@@ -3,44 +3,86 @@ package ntske
...
+		Enabled    bool
+		RemoteAddr udp.UDPAddr
+		LocalAddr  udp.UDPAddr
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/fetcher.go]: Done

File: net/ntske/fetcher.go [hunk 52]
```diff
@@ -3,44 +3,86 @@ package ntske
 import (
 	"crypto/tls"
 	"errors"
+	"log"
 	"net"
 
 	"go.uber.org/zap"
+
+	"github.com/quic-go/quic-go"
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/fetcher.go]: Done

File: net/ntske/ntske_scion.go [hunk 53]
```diff
@@ -0,0 +1,99 @@
...
+
+	_, err = stream.Write(buf.Bytes())
+	if err != nil {
+		return err
+	}
+	quic.SendStream(stream).Close()
+	reader := bufio.NewReader(stream)
+
+	// Wait for response
+	err = Read(log, reader, data)
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go]: Done

File: net/ntske/ntske_scion.go [hunk 54]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/quic-go/quic-go"
+	"go.uber.org/zap"
+
+	"github.com/scionproto/scion/pkg/daemon"
+
+	"example.com/scion-time/net/scion"
+	"example.com/scion-time/net/udp"
+)
+
+func newDaemonConnector(log *zap.Logger, ctx context.Context, daemonAddr string) daemon.Connector {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply net/ntske/ntske_scion.go]: Done

**Review by marcfrei — APPROVED** [review]

**Review by marcfrei — APPROVED** [review]

**Pull request merged**